### DISCUSSION TITLE
x-anchor dynamic $ref or id

### DISCUSSION THREAD

**Discussion by nam-co** [alpinejs/alpine#4079] [1. Help] [answered]
Hello, is there a way to update the anchor point, for example if you have 10 buttons that show the same x-anchor popup, instead of duplicating the code, is there a way to update x-anchor="$refs.updateThis" or x-anchor="document.getElementById(changeMe)"
Sample:
https://codepen.io/nam_co/pen/xxewXVQ

Appreciate any help

As of early 2024, this is probably the best way, hopefully dynamic targets will come soon, and AUTO position pls, thanks @docdunning  for all the help

**Comment by nam-co** [comment]
Maybe @ekwoka has an idea

**Comment by docdunning** [comment, marked as answer]
The anchor can only be changed when the popup is not being rendered.

You can achieve this by using `x-if` on a template around the popup, instead of `x-show`.  The `x-anchor` position is then recalculated when the template becomes "active" and the popup is rendered.  

I understand that you want to avoid duplication, but I'm not sure how you would use this in reality. 

```
<div 
    x-data="{
      open: false,
      anchor: 'button1',
      changeAnchor() {
        this.open = false
        this.anchor = this.anchor=='button1' ? 'button2' : 'button1'
      }
    }">
    <p><button x-ref="button1" @click="open = ! open">Toggle 1</button></p>
    <p>
      <button @click="changeAnchor()">Change anchor</button>
      <span x-text="anchor"></span>
    </p>  
    <p><button x-ref="button2" @click="open = ! open">Toggle 2</button></p>
    <template x-if="open">
      <div x-anchor="$refs[anchor]">
        Popup content anchored to <span x-text="anchor"></span>
      </div>
    </template>
</div>

```

**Reply by nam-co** [reply]
thanks @docdunning thank you for your idea, what do you think of https://codepen.io/nam_co/pen/xxewXVQ?editors=1010

**Reply by docdunning** [reply]
Yep, that works.

Here's an alternative, which doesn't use the `x-anchor` plugin.  I guess it depends on how much of the plugin's functionality you need.  This example just uses JS to move the popup element in the DOM, so that it follows the button that was clicked.

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
  <style>
    .container {
      position: relative;
    }
    #popup {
      position: absolute;
      z-index: 1;
      top: 8px;
      left: 8px;
      padding: 10px;
      background-color:#333;
      color: #fff
    }
  </style>
</head>
<body>
  <div 
    x-data="{
      open: false,
      addPopup() {
        this.$el.after(this.$refs.popup)
        this.open = true
      },
      closePopup() {
        this.open = false
      }
    }">
    <div class="container">
      <button @click="addPopup()">Popup 1</button>
      <div id="popup" x-ref="popup" x-show="open" @click.outside="closePopup()">Popup</div>
    </div>
    <div class="container">
      <button @click="addPopup()">Popup 2</button>
    </div>
  </div>
</body>
</html>
```

**Reply by nam-co** [reply]
@docdunning thank you very much for the code and the ideas, I will test it
